modules/runner/src/main/java/org/apache/ignite/internal/storage/DistributedConfigurationStorage.java

```diff
@@ -49,52 +72,154 @@ public DistributedConfigurationStorage(MetaStorageManager metaStorageMgr) {
+
             if (entry.getValue() != null)
-                map.put(entry.getKey(), entry.getValue());
+                // TODO: investigate overhead when deserialize int, long, double, boolean, string, arrays of above
+                operations.add(Operations.put(key, ByteUtils.toBytes(entry.getValue())));
             else
-                map.remove(entry.getKey());
-        }
+                operations.add(Operations.remove(key));
 
-        this.version.incrementAndGet();
+            failures.add(Operations.noop());
+        }
 
-        listeners.forEach(listener -> listener.onEntriesChanged(new Data(newValues, this.version.get(), 0)));
+        operations.add(Operations.put(masterKey, ByteUtils.longToBytes(sentVersion)));
 
-        return CompletableFuture.completedFuture(true);
+        return metaStorageMgr.invoke(Conditions.key(masterKey).revision().eq(version.get()), operations, failures);
     }
 
     /** {@inheritDoc} */
-    @Override public void addListener(ConfigurationStorageListener listener) {
+    @Override public synchronized void addListener(ConfigurationStorageListener listener) {
         listeners.add(listener);
+
+        if (watchId == null) {
+            watchId = metaStorageMgr.registerWatchByPrefix(masterKey, new WatchListener() {
+                @Override public boolean onUpdate(@NotNull Iterable<WatchEvent> events) {
+                    HashMap<String, Serializable> data = new HashMap<>();
+
+                    long maxRevision = 0L;
+
+                    Entry entryForMasterKey = null;
+
+                    for (WatchEvent event : events) {
+                        Entry e = event.newEntry();
+
+                        if (!e.key().equals(masterKey)) {
+                            data.put(e.key().toString().replaceFirst(DISTRIBUTED_PREFIX + ".", ""),
+                                (Serializable)ByteUtils.fromBytes(e.value()));
+
+                            if (maxRevision < e.revision())
+                                maxRevision = e.revision();
+                        } else
+                            entryForMasterKey = e;
+                    }
+
+                    // Contract of metastorage ensures that all updates of one revision will come in one batch.
+                    // Also masterKey should be updated every time when we update cfg.
+                    // That means that masterKey update must be included in the batch.
+                    assert entryForMasterKey != null;
+
+                    assert maxRevision == entryForMasterKey.revision();
+
+                    assert maxRevision >= version.get();
+
+                    long finalMaxRevision = maxRevision;
+
+                    listeners.forEach(listener -> listener.onEntriesChanged(new Data(data, finalMaxRevision)));
+
+                    return true;
+                }
+
+                @Override public void onError(@NotNull Throwable e) {
+                    LOG.error("Metastorage listener issue", e);
```

Comment:
According to metaStorage onError contract
`The method will be called in case of an error occurred. The listener and corresponding watch will be unregistered.` 
So please consider redeploying watches, otherwise you will never receive watch notifications.

Reply:
added todo with ticket